Parallelize checkout_create_the_new for ntfs perf gains

implausible opened the pull request:
So we noticed that checkout performance on Windows boxes is particularly slow in comparison to OSX / Linux boxes. Windows checkout operations can run 30x slower than an equivalent checkout operation in a windows box. After running some profiling with Visual studio we tracked the cause of the performance to the method `checkout_create_the_new`, which internally opens files for writes internally. In fact, the bottle neck ended up being the time windows is taking to talk with the file system. This seems crazy!

So, following the suggestons of @ethomson, we tried performing some checkouts using an EXT4 partition via ext2fsd on windows, and we actually dropped from 30 second checkouts to 10 second checkouts! This made it very clear that NTFS is getting in the way of fast checkouts in Windows.

So, on another suggestion of @ethomson, we dug in and turned checkout_create_the_new into a thread pool. We found that the number of threads per core should be roughly the same - it appears that the slowdown of NTFS is primarily the CPU overhead of compression / security checks and anything else NTFS does for you on every open/write/read call. So we didn't gain anything statistically significant by increasing the number of threads per core.

That said, our metrics show that with this added into libgit2, windows checkouts perform 30-40% faster than the CLI. Very significant improvements!

implausible commented:
Just wanted to come back and add some more data. I did some perf testing on an 8 core machine, same checkouts. 60-70% perf improvement there. :D

implausible commented:
So we were doing some testing here and super duper finally discovered the slowdown on Windows machines. Windows Defender is the culprit. When we disable Windows Defender, performance on Windows machines runs near the same speed as a Linux or OS/X box. Now that said, Defender or an equivalent is often enabled in corporate environments, so this optimization is still a good thing. We still see crazy perf gains for those with active virus protection...

That said, since we've found the real culprit and it's not actually NTFS... We're not just optimizing the Windows pathway anymore, we're really just optimizing checkouts in general. So I'm removing the flag that restricts this behavior for Windows only, because we could still see perf increases in other environments.

ethomson commented:
I'm curious to know what kind of behavior changes you see in non-Windows environments?

implausible commented:
I just found I needed to share the index mutex over the `git_filter_list__load_ext` and ditch the filter_mutex. It seems somewhere deep down in `git_filter_list__load_ext` we end up touching the index, and I managed to get a crash there.

implausible commented:
For Linux we see about a 10-20% increase in speed from previous versions of libgit2 using the same test branches.

Libgit2 is still slower than the CLI on Linux. Even after threading, CLI checkout performance is outstanding on Linux. After threading we are 100-200ms slower than the CLI, and about 30-60% slower than the CLI, but everything post threading in our tests with a large checkout operation is under 1 second. So there is more room for us to optimize the checkout code, but it would probably be chasing some minor perf gains in the grand scheme of things. Most of those performance tweaks would be more impactful on Windows machines than Linux.

implausible commented:
I wonder if instead of locking the index on the whole filter list if we could utilize the `git_attr_session` to pass the index mutex down to where the index is actually being used. @ethomson thoughts?

implausible commented:
This seems to be failing because of the bitbucket issue in PR https://github.com/libgit2/libgit2/pull/4584

implausible commented:
We've been running this in GK through nodegit for almost 9 months now without seeing any craziness.

implausible commented:
Force pushed to fix some conflicts from the git_buf_free -> git_buf_dispose change. Also, we'll need to continue cherrypicking this in NodeGit for the time being.

@ethomson @tiennou @pks-t @carlosmn  I know I've been bugging you a lot lately this week, but might I bug you one more and ask if anyone has time for this change :smile:

tiennou commented:
(accesses to `git_pool` seems to also be a concern, as I don't think those are threadsafe)

implausible commented:
Thank you for the review @tiennou! I apologize for the lack of commits a story telling mechanism. Now that this is receiving attention, again. When I am reconciling your review, I will try to rebuild the patches to help make this review more digestable.

ethomson commented:
> Thank you for the review @tiennou! I apologize for the lack of commits a story telling mechanism. Now that this is receiving attention, again. When I am reconciling your review, I will try to rebuild the patches to help make this review more digestable.

Another option is to take bits of this PR and merge them opportunistically, as it makes sense.  I could imagine bringing in some refactoring, etc.  This is sort of what we did for the original merge PR...  I think that would be pretty straightforward if that was a direction that you wanted to go instead.

Just throwing that out there; I'm happy either way.

implausible commented:
So while these numbers are not checkout in isolation, these are some numbers that the GitKraken team took after multithreading checkout in GitKraken.

In these documents, for reference, `2.4 GK` refers to checkout without threading and `2.5 GK RC3` refers to checkout operations with threading.

This first set of numbers comes from Windows machines.
https://docs.google.com/spreadsheets/d/1ciR-llAkKS0zcZYlOBzR8BLJzRHU5y5qu7KDym6REo0/edit?usp=sharing

This second set of numbers is performing the same checkout in GitKraken but on a Linux machine.
https://docs.google.com/spreadsheets/d/1RDDJFyWMVVSoHXbsRhNoNuVEO_4iZA31twGEoyZeJ_A/edit?usp=sharing

Unfortunately, the repository we used is an old, large, private repo for the Axosoft product line, so I cannot share the repo that helped to create these numbers. You'll have to trust that this particular checkout operation is very large :).

I do not have Mac OS numbers at this moment, so it probably would be good to redo these numbers anyway in a more isolated form later.

tiennou commented:
I've put out my patchset here : https://github.com/libgit2/libgit2/compare/master...tiennou:feature/threaded-checkout-clean.

It's not quite pretty enough, but I think it makes what's going on more isolated (at least to me 😉).

implausible commented:
@tiennou I think we need to be careful in `wd_item_is_removable` to dispose the fullPath regardless of the success or failure of `build_target_fullpath` in your patch set. Perhaps we should be failing that method of `git_buf_set` fails as well?

implausible commented:
I've taken that patchset. Thanks for doing that.

implausible commented on src/checkout.c:
```diff
@@ -1875,6 +1920,33 @@ static int checkout_deferred_remove(git_repository *repo, const char *path)
+		if (error < 0)
+			return error;
+
+		data->completed_steps++;
```
I feel the struggle here as well. It's possible we could create a more complicated contract for `checkout_create_perform` that both the single and parallel `checkout_create_the_new` need to handle.

tiennou replied:
I actually have a [WIP patch](https://github.com/tiennou/libgit2/compare/feature/threaded-checkout-clean...tiennou:feature/threaded-checkout-wip) for that, but I broke the test case (something about skipped/errored pairs being miscounted). Though I'm not sure I like it as it is…

implausible commented on src/checkout.c:
```diff
@@ -1527,18 +1547,26 @@ static int blob_content_to_file(
 
 	filter_opts.attr_session = &data->attr_session;
-	filter_opts.temp_buf = &data->tmp;
+	filter_opts.temp_buf = &temp_buf;
```
I wonder if that is why I needed to build a mutex around the filters?

tiennou replied:
I… think it does, yes. I've singled that out because there are some other users of that `data->tmp` buffer and thought they might clobber as well, but after more time reading the code, it seems to only serve as a temporary buffer for the filter application (ie. it shouldn' t share its lifetime). So it makes sense IMHO to just use a stack-allocated buf at the right level.

The need for locking is still unconfirmed though 😉.

tiennou commented on src/checkout.c:
```diff
@@ -1875,7 +1928,34 @@ static int checkout_deferred_remove(git_repository *repo, const char *path)
+		/* this had a blocker directory that should only be removed iff
+		 * all of the contents of the directory were safely removed
+		 */
+		if ((error = checkout_deferred_remove(data->repo, delta->old_file.path)) < 0)
```
Just for kicks, the body of this function is `#ifdef`ed out to do nothing 😂.

implausible replied:
:joy:

implausible replied:
That block is 6 years old. I'm assuming it's not going to change anytime soon. I think I'll remove that then.

implausible commented:
Tests failed due to connection issues.

implausible commented:
@ethomson @tiennou Ok so I ran into some annoying conflicts early last week with this branch. I decided I would take a stab at cleaning up some of the items that needed work for this PR.

The major change is that we're based on https://github.com/libgit2/libgit2/pull/5602, because it allows us to easily address some of @ethomson's concerns he had expressed to me over performance regressions in the single threaded case due to over allocations of the tmp / target_path buffers.

The other changes are:
- perfdata now just uses a simple lock before writes
- completed_steps needed a lock to make it thread safe
- I've added a separate mutex for the `git_filter_list__load_ext`, as I am reasonably certain that it is the attribute cache on the repository that makes that method not threadsafe.

I've also tried hard to keep each commit easily digestible. Swapping the temporary buffers to use TLS has immeasurably reduced the footprint of what needs to be reviewed for this PR.

Now on to what I have not implemented yet.. I have not gone through and added this as configurable in an option (whether on checkout_options or in libgit2_opts), my hold up here is on testing. Checkout is such an integral part of many other things you can do in libgit2 that I want anything that could potentially run checkout to run in threaded mode during testing of the library... I was also thinking that whether you thread checkout or not seems like a library option rather than an individual checkout option.

So what I would like to do is just duplicate the offline checkout tests and at least one of the online tests that involves cloning and turn on checkout threading on those runs.. I also think, because of the nature of threading, we should require the offline checkout tests to run 4 times per test cycle to hopefully catch timing errors that could crop up someday.

I would really appreciate it if you guys could get back to me soon on this one, I would like to add the option and get this merged in asap - I would prefer not to deal with another merge conflict from this area.

implausible commented:
> especially since there are many places where we might be stepping all over our own threading expectations (ie. you should never share git_repository instances between threads, so I shudder each time I see that checkout_data parameter), and auditing for that is… complicated (as I have no experience with our checkout implementation).

@tiennou I think in general, the reason the pool is working correctly is that we are locking around its use in `mkpath2file -> checkout_mkdir`. I could alleviate your fears by moving the pool into TLS.

The only other place that I see that is potentially dangerous in an audit (while I'm fresh on this again) is access to the ODB in `checkout_write_content` where we are performing a lookup of a blob there, as well as freeing it. Should we be locking on blob lookup and free?

implausible commented:
I know I've still got ergonomics to do on this, but seeing the fuzzer agree with this branch makes me feel better about it.
